### PR TITLE
fix list parsing

### PR DESCRIPTION
**Pull request by nicolas-martin**
There was a bug where if an list contained only 1 item, it would create an option instead of keeping it as a list. This bug can be replicated by setting [dnsmasq](https://openwrt.org/docs/guide-user/base-system/dhcp) server as option `option server '127.0.0.1#54'` and observing the that config in `/var/etc/dnsmasq.conf.xxx` does not contain the server field. Unlike when setting `list server '127.0.0.1#54'`.

- I exported `ItemType` as well as `ItemOption` and `ItemList
- Added `Type ItemType` field to `option` struct
- Created a new `SetType` which takes an `ItemType` in order to create a list.
- `Set` still works as before and now calls `SetType` with `ItemOption` param.

Looks like a lot of changes. But it's was mostly refactoring.

### CODE REVIEW

**Comment by dmke** [comment]
Thanks for catching this. Here are my thoughts:

I don't like to export only *some* bits of the lexer internals. I would instead prefer to introduce a distinct type for the option type, something akin to:

```go
type OptionType int
const (
    TypeOption OptionType = iota // option is not a list
    TypeList                     // option is a list
)

func (ot OptionType) itemType() itemType {
    switch ot {
    case TypeOption:
        return itemOption
    case TypeList:
        return itemList
    default:
        panic("not reached")
    }
}

func (t *tree) SetType(config, section, option string, typ OptionType, values ...string) bool {
    ...
    sec.Add(newOption(option, typ.itemType(), values...))
    ...
}
```

That way you can keep the diff smaller. It also catches calls with the wrong option type early: `SetType(cfg, sec, opt, 0, vs...)`.

For the tests, I also suggest to implement the JSON `Marshaler` and `Unmarshaler` interfaces. That would improve the readability and maintainability of the test fixtures (`type": 5` vs `"type": "option"`):

```go
// MarshalJSON implements encoding/json.Marshaler.
func (ot OptionType) MarshalJSON() ([]byte, error) {
    switch ot {
    case TypeOption:
        return []byte("option"), nil
    case TypeList:
        return []byte("list"), nil
    default:
        return nil, errUnknownOptionType
    }
}

// UnmarshalJSON implements encoding/json.Unmarshaler.
func (ot *OptionType) UnmarshalJSON(b []byte) error {
    if len(b) == 0 || bytes.Equal(b, []byte(null)) || bytes.Equal(b, []byte("option")) {
        *ot = TypeOption
        return nil
    }
    if bytes.Equal(b, []byte("list")) {
        *ot = TypeList
        return nil
    }
    return errUnknownOptionType
}
```

**Comment by nicolas-martin** [comment]
Great suggestions. One question though, do you want the [option](https://github.com/digineo/go-uci/blob/master/types.go#L271-L274) struct to hold the internal `itemType` or `OptionType`? From your suggestion
```
sec.Add(newOption(option, typ.itemType(), values...))
```
Looks like it holds `itemType`. But from your other marshalling suggestion it looks it's holding `OptionType`

File: uci.go [hunk 1]
```diff
@@ -252,12 +255,16 @@ func (t *tree) Set(config, section, option string, values ...string) bool {
...
-		sec.Add(newOption(option, values...))
+		sec.Add(newOption(option, optionType, values...))
 	}
 	cfg.tainted = true
 	return true
 }
 
+func (t *tree) Set(config, section, option string, values ...string) bool {
+	return t.SetType(config, section, option, TypeOption, values...)
+}
```
Comment (dmke) [inline uci.go:266]: Hm, maybe the `OptionType` here should depend on `len(values)`? This was previously decided in `types.go`:

```go
for _, opt := range sec.Options {
	switch l := len(opt.Values); {
	case l == 1:
		// option
	case l > 1:
		// list
	}
}
```

Additionally, I'm in favour of deprecating `Set`:

```suggestion
// Set will either set an option or a list at the given configuration section.
// The option/list decision is based on the number of values.
// Deprecated: Use SetType to control the type.
func (t *tree) Set(config, section, option string, values ...string) bool {
	if len(values) > 1 {
		return t.SetType(config, section, option, TypeList, values...)
	}
	return t.SetType(config, section, option, TypeOption, values...)
}
```

**Review by dmke — CHANGES_REQUESTED** [review]
I like what I'm seeing, but I have a few requests for changes/comments.

File: uci.go (hunk 1 above)
Comment (dmke) [reply uci.go:266]: I just noticed: the comment should go into the `Tree` interface.

Comment (nicolas-martin) [reply uci.go:266]: This was part of the logic that was causing a failure. It's totally valid for a list to only have 1 value. List and option are not interchangeable.

**Review by dmke — APPROVED** [review]

**Comment by dmke** [comment]
Thank you for this PR. I will merge this tomorrow together with some other changes.

**Pull request closed without merging**